elements/pl-dataframe/pl-dataframe.py

```diff
@@ -0,0 +1,164 @@
+from enum import Enum
+from typing import cast
+
+import chevron
+import lxml.html
+import pandas as pd
+import prairielearn as pl
+from typing_extensions import assert_never
+
+
+class DisplayLanguage(Enum):
+    PYTHON = 1
+    R = 2
+
+
+SHOW_HEADER_DEFAULT = True
+SHOW_INDEX_DEFAULT = True
+SHOW_DIMENSIONS_DEFAULT = True
+DISPLAY_LANGUAGE_DEFAULT = DisplayLanguage.PYTHON
+SHOW_DTYPE_DEFAULT = False
+NUM_DIGITS_DEFAULT = None
+SHOW_PYTHON_DEFAULT = True
+
+
+def convert_pandas_dtype_to_r(s: pd.Series) -> str:
+    # Force series to avoid odd element-wise output
+    s.dtype
+
+    if pd.api.types.is_float_dtype(s):
+        return "numeric"
+    elif pd.api.types.is_integer_dtype(s):
+        return "integer"
+    elif pd.api.types.is_object_dtype(s) or pd.api.types.is_string_dtype(s):
+        return "character"
+    elif pd.api.types.is_categorical_dtype(s):
+        # Check if ordered
+        if s.cat.ordered:
+            return "ordered factor"
+        return "factor"
+    elif pd.api.types.is_bool_dtype(s):
+        return "logical"
+    elif pd.api.types.is_complex_dtype(s):
+        return "complex"
+    elif pd.api.types.is_datetime64_any_dtype(s):
+        return "POSIXct"
+    elif pd.api.types.is_timedelta64_dtype(s) or pd.api.types.is_period_dtype(s):
+        return "Not supported"
+
+    return "Unknown"
+
+
+def get_pandas_dtype(s: pd.Series) -> str:
+    return s.dtype
+
+
+def prepare(element_html: str, data: pl.QuestionData) -> None:
+    element = lxml.html.fragment_fromstring(element_html)
+    pl.check_attribs(
+        element,
+        required_attribs=["params-name"],
+        optional_attribs=[
+            "show-index",
+            "show-header",
+            "show-dimensions",
+            "show-dtype",
+            "show-python",
+            "display-language",
+            "digits",
+        ],
+    )
+
+
+def render(element_html: str, data: pl.QuestionData) -> str:
+    element = lxml.html.fragment_fromstring(element_html)
+    # TODO should there be a different display varname?
+    varname = pl.get_string_attrib(element, "params-name")
+    show_index = pl.get_boolean_attrib(element, "show-index", SHOW_INDEX_DEFAULT)
+    show_header = pl.get_boolean_attrib(element, "show-header", SHOW_HEADER_DEFAULT)
+    show_python = pl.get_boolean_attrib(element, "show-python", SHOW_PYTHON_DEFAULT)
+    show_dimensions = pl.get_boolean_attrib(
+        element, "show-dimensions", SHOW_DIMENSIONS_DEFAULT
+    )
+    show_dtype = pl.get_boolean_attrib(element, "show-dtype", SHOW_DTYPE_DEFAULT)
+
+    display_language = pl.get_enum_attrib(
+        element, "display-language", DisplayLanguage, DISPLAY_LANGUAGE_DEFAULT
+    )
+
+    num_digits = pl.get_integer_attrib(element, "digits", NUM_DIGITS_DEFAULT)
+
+    if varname not in data["params"]:
+        raise KeyError(
+            f'Could not find {varname} in params. Please make sure to set params-name="{varname}" in the element.'
```

Comment:
Not sure if this error message is useful. If we reach this point in the code, then `params-name` *must* have been set in the HTML for this element.

Reply:
That's a good point, will remove the extra text.